rfcs/text/0000-simple-updates.md

```diff
@@ -0,0 +1,159 @@
+***Old update process*** - the current update process which involves adding the existing project to an exclusion list and re-ingesting the entire project - with updates - from scratch as a new submission  
+***Data wrangler*** - member of the DCP team who interacts with data contributors to collect metadata updates and additions and drives the process of submitting them to the DCP  
+***Experimental design representation*** - the connected set of entities which the DCP uses to represent the experimental design of a project  
+***Bundles*** - The data store representation of a logical set of data and metadata. There are two types of bundles input and analysis. Input bundles represent a single set of experimental data such as a lane or plex of sequencing and are defined by ingest as a biomaterial to file transition, connected by an assay process and all entities upwards in the experimental design from that point. Analysis bundles are all the analysis results from a single run of an analysis pipeline on a single input bundle and are defined by ingest as a file to file transition as defined by an analysis process and all entities upwards in the experimental design from that point.
+
+## User Stories
+
+As a data contributor, I would like to add the details of a scientific publication to my submission to the DCP to broaden knowledge of the work.
+
+As a data contributor, I would like to correct an error I have made in my submission to ensure that the correct metadata is presented to consumers of my data.
+
+As a data wrangler, I would like to update organ system metadata to reflect an improved ontology annotation for this particular organ.  
+
+Read more on the general user experience background for this work: [User Experience Problem Space - Top task “Update my project”](https://docs.google.com/document/d/1FP6kJa3f9NkOHTLqmkNYztls7soZLgeGefxdgR81ONE/edit)
+
+## Scientific "guardrails"
+
+The proposed design has consequences for end users where different types of updates will have different impact on how data is visible in the DCP. Updates and additions which are in scope for this solution will have provanance trails which could be displayed to users but other updates such as those which change the experimental design representation or require reanalysis will still use the old update process and won't have any provanance reflected inside the DCP. 
+
+We propose the UX team work with the community and the DCP Oversight Committee, building on the existing research to understand our data consumers assumptions around updates and to guide how we display updates to consumers in the data browser and other consumer services both now and in the future.  
+
+## Detailed Design
+
+### User Interactions
+
+These are the steps that a wrangler will follow to perform a primary metadata update.
+
+![](../images/0000-simple-updates-ingest-flow.png)
+
+* **Step 1 - Wrangler finds submission**
```

Comment:
This RFC is concentrating on the requirements as they exist now, which is to allow wranglers to make updates. There are no users of this functionality in Q2 apart from the ingest UI.

Hence, I think it is better to keep these APIs private (unpublished) until other components need to use them. At that point, we can revise them in the light of our experience before RFCing and publishing them. Increasing out public API surface at this point will make it more difficult to do future revisions without any obvious benefit. It's also the case that these details are not yet pinned down at the low level as ingest design for updates is complex. Fully forming this API to a publishable level will take extra time.

That's my opinion. @lauraclarke, @morrisonnorman please could you comment.

Reply:
The API could be marked as developmental/provisional, but the API needs to be there ("public" in the sense of available for applications other than the first-party UI to call, not "open to all public at large"). The users of this functionality are/will be the wranglers, who can't develop automation for their work without a documented API. I think at this stage it's very reasonable for that API to change quickly in breaking ways.

Reply:
@kislyuk I agree with Justin at this early point in the design this introduces some risk especially as during Q2 there are no expected programmatic users.

This is because wranglers only interact with the service via the UI. Currently, we only have one user of the programmatic service, the pipelines team and as they aren't expecting to submit updates this quarter.  

This means avoiding specifying the endpoints now mitigating the risk of the design needing to change. If we were immediately expecting programmatic users and for them to comment if the endpoints were reasonable it would be different but that isn't the position we are in.

Reply:
In my opinion it is important to always define an API first, and then define a UI (and possibly some UI-specific "value added" service functionality) on top of it. This allows one not to get into uncomfortable situations where the reason no automation exists is that no tools were available for automation, or where an API has to be retrofitted on top of a system that organically evolved to fuse API and UI design decisions together. I think it's also important that we get away from having the spreadsheet-centric paradigm of data handling be the lowest tier of the ingest service data interchange model. Spreadsheets are a fine way to work with metadata, but the underlying system and API should be able to accept the constituent metadata in its raw form as arrays or associative arrays of strings.

The ingest service does have a commitment to define an API as part of its charter, and I think an API sketch should be the starting point of any RFC of this type. We should get into the habit of centering our discussions around that.

With all that said, I understand the time and resource constraints that prompted this RFC's scope restrictions. To clarify the intent, I propose adding a new section to this RFC, stating that API design for this feature is postponed but forthcoming. Here is my proposed wording - please feel free to change it as you see fit:

--------------

## Detailed Design
### Metadata Updates API

The Ingest Service API provides a key capability for data wranglers to provide wrangling portals (interfaces) appropriate for their specific user or data modality needs, as well as for automation of bulk/batch data submission, data QC and status monitoring. The Ingest Service API also provides programmatic access for the analysis pipeline execution manager (however, metadata updates from analysis pipelines are beyond the scope of this RFC).

The Ingest Service API and its attendant developer resources ([documentation](https://github.com/HumanCellAtlas/ingest-core) and [API client](https://github.com/HumanCellAtlas/ingest-client)) will be updated to accommodate metadata update functionality as described in this RFC. The design of these API updates is beyond the scope of this RFC, and will be addressed in a future RFC.

--------------

cc @claymfischer @cjvillar